feat: support preloading also on legacy bundle

Tal500 opened the pull request:
### Description
Support preloading also on legacy bundle.

A test case is included for a utility function I'm using.

closes #9902 and even more (since it's also preload JS scripts)
fixes #2062
closes #5901
fixes #10285

### Additional context

As mentioned in #9902 and #10285, by solving the issue with this PR, it makes the legacy build support for SvelteKit I'm working on(sveltejs/kit#6265) to load correctly the CSS files when navigating(and more correctly, on preloading) from one page to the other.

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [X] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [X] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

Tal500 committed:
feat: support preloading also in legacy bundle

Tal500 committed:
chore: describe the `parseImportsSystemJS` function better

Tal500 committed:
test: ignoring code file for testing in prettier

Tal500 committed:
chore: adapt the code to work also in Node 12

Tal500 committed:
chore: test a change in the code

Tal500 committed:
revert: chore: test a change in the code

Tal500 committed:
chore: merge branch 'preload-on-system-format' of github.com:Tal500/vite into preload-on-system-format

Tal500 committed:
chore: fix revert issues

Tal500 commented:
Help: Don't know why vue legacy tests are failing.

sapphi-red commented:
It seems this image is not being loaded when doing `pnpm build && pnpm preview` in `playground/vue-legacy`.
https://github.com/vitejs/vite/blob/4158b98c8fc30070967de1176f4b1c252d275386/playground/vue-legacy/module.vue#L9

Tal500 committed:
fix: fix the css regexp

Tal500 commented:
> It seems this image is not being loaded when doing `pnpm build && pnpm preview` in `playground/vue-legacy`.
> 
> https://github.com/vitejs/vite/blob/4158b98c8fc30070967de1176f4b1c252d275386/playground/vue-legacy/module.vue#L9

Tnx! I had realized that the root cause for the failure is because I'm fighting with Vite way to inline the CSS on legacy.
But the immediate question now is, why?
I have commented this in the original issue I had opened:
https://github.com/vitejs/vite/issues/9902#issuecomment-1236194069

Tal500 committed:
fix: return the format to `module.import` in regex

Tal500 committed:
fix: and simplify the RE in `parseImportsSystemJS`

Tal500 committed:
test: more test snippets for `parseImportsSystemJS`

Tal500 committed:
fix: safer detection of import regexp for empty css's

Tal500 committed:
chore: omit the obsolete css inlining on legacy

Tal500 committed:
test: fix line ending issues while reading, and update the snapshot

Tal500 committed:
test: (hopefully) fix line ending issues

Tal500 committed:
test: and code files for testing to lint ignore

Tal500 committed:
feat: better check on seeing whether the asset path was already loaded

This is done by comparing short pathes to the full ones.

Tal500 committed:
fix: a much safer way to load CSS on legacy, while try to save performance by caching.

Tal500 committed:
chore: fixing the `@ts-ignore` location (fighting with lint)

Tal500 committed:
fix: linting issues

Tal500 committed:
feat: conditional waiting in preloading

Tal500 committed:
fix: make sure to have the full dep path

Tal500 committed:
fix: try to break also when paths are simply equal

and improve typing

Tal500 committed:
chore: remove full dep equlity optimization for checking tests

sapphi-red commented:
I tested with Chrome 16 and it worked greatly. 👍 (when minify is disabled)

Tal500 commented on packages/vite/src/node/plugins/importAnalysisBuild.ts:
```diff
@@ -73,30 +72,97 @@ function preload(
+        }
+      } else {
+        if (isCss) {
+          return new Promise<void>(preloadInImg).then(loadLinkAndWait)
```
> Maybe we could simplify this to `XHR -> add link` from `load img -> add link -> XHR`?

I'm not sure about browsers behavior.
There's some feeling in my stomach that if you load a content as image, it is "much more cached" in some sense (e.g. having two images with the same source in the same page, will probably look always the same to each other, even if their source isn't static).

As you can see in https://github.com/vitejs/vite/blob/e72c0905bec8b0540fa41622956d26028beed04e/packages/vite/src/node/plugins/importAnalysisBuild.ts#L127-L138, the only reason I'm calling XHR **is just for the ability to know when there is an error on loading the content**.
Shell we just don't care about reporting error on loading in legacy? This will simplify the flow to Img -> Style

Tal500 replied:
By "much more cached" - I mean that it seems like there is "cache" and there is "page resources".
When loading to image, it's get loaded to both of them, and on XHR, it seems to be loaded only to the cache.

If we load it to the page resources, it feels like the CSS will be immediately available after, unlike restoring it from the cache.

Tal500 replied:
See now the new preload function, I had stripped off the XHR request, resulting in removing the ability to detect an error on fetching the CSS data when on legacy (I don't think it's so important, and makes things smoother):
https://github.com/vitejs/vite/blob/6dbe67656040aab79e9c9c0e32944f56c0856ea2/packages/vite/src/node/plugins/importAnalysisBuild.ts#L58-L147

Tal500 commented on packages/vite/src/node/plugins/importAnalysisBuild.ts:
```diff
@@ -73,30 +72,97 @@ function preload(
+        if (isCss) {
+          return new Promise<void>(preloadInImg).then(loadLinkAndWait)
+        } else {
+          preloadInImg(() => {})
```
OK, will be fixed

Tal500 commented on packages/vite/src/node/plugins/importAnalysisBuild.ts:
```diff
@@ -379,7 +445,11 @@ export function buildImportAnalysisPlugin(config: ResolvedConfig): Plugin {
-            imports = parseImports(code)[0].filter((i) => i.d > -1)
+            imports = (
+              format === 'system'
+                ? parseImportsSystemJS(code)
```
OK, so I got to first analyze the parameter name in System.register, and then use this name in expressions.

Tal500 replied:
Additionally, the same issue should be also here:
https://github.com/vitejs/vite/blob/6dbe67656040aab79e9c9c0e32944f56c0856ea2/packages/vite/src/node/plugins/css.ts#L586

And yes, I know, I forgot it should be started with `\\b` to mark a non-word boundary, will be fixed.

I'll make a function in `utils.ts` that detect SystemJS register parameters, and use it for both plugins I had modified.

Tal500 committed:
chore: simplify the code by removing the xhr request and thanks to the module preload polyfill

Note: Removing the XHR request means that there will be no reported error on legacy if the CSS resource isn't found.

Tal500 committed:
chore: minor style improvement

Tal500 committed:
fix: add and use `analyzeSystemRegisteration` to handle parameter name change on minimization

added and reordered additional test to `analyzeSystemRegisteration`

Tal500 committed:
build: change the dir to ignore on eslint

Tal500 commented:
Test now, it's working on my machine also after minimization, including testing.

Note: As said, I had stripped out the XHR call. This results with the side effect that if the network resource(the CSS file) isn't available, then no exception will be raised on legacy.

If this is crucial I can return this extra call.

Note2: I had modified a little bit the directory structure of the testing, to include the code samples. Hope it's fine

Tal500 committed:
chore: simplify the code

bgoscinski commented on packages/vite/src/node/utils.ts:
```diff
@@ -1150,3 +1150,44 @@ export const isNonDriveRelativeAbsolutePath = (p: string): boolean => {
+/**
+ * Analyze the SystemJS module registration `System.register` in the code.
+ */
+export function analyzeSystemRegisteration(code: string):
```
```suggestion
export function analyzeSystemRegistration(code: string):
```

bgoscinski commented on packages/vite/src/node/__tests__/utils.spec.ts:
```diff
@@ -177,3 +196,13 @@ describe('asyncFlatten', () => {
   })
 })
+
+describe('analyzeSystemRegisteration', () => {
```
```suggestion
describe('analyzeSystemRegistration', () => {
```

bgoscinski commented on packages/vite/src/node/plugins/importAnalysisBuild.ts:
```diff
@@ -73,30 +73,66 @@ function preload(
       if (dep in seen) return
       // @ts-ignore
       seen[dep] = true
+      const seperatorIdx = dep.lastIndexOf('/')
```
```suggestion
      const separatorIdx = dep.lastIndexOf('/')
```

bgoscinski commented on packages/vite/src/node/plugins/importAnalysisBuild.ts:
```diff
@@ -73,30 +73,66 @@ function preload(
+        const currentPath = possibleLinks[i].href
+        if (
+          currentPath === dep ||
+          new URL(currentPath, importerUrl).href === dep
```
Unless I'm missing something I think that line 88 can be safely removed. Here's my reasoning:

`possibleLinks[i].href` is an absolute URL already because this expression accesses the href _property_ (not an _attribute_) and browsers must normalize all URL containing _properties_ to be absolute. See [here](https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#reflecting-content-attributes-in-idl-attributes:idl-domstring-5) under "If the content attribute is defined to contain a URL". Because of that `currentPath` also must be absolute. Doing `new URL(AN_ABSOLUTE_URL, whatever).href` returns `AN_ABSOLUTE_URL` so 

```
// original code
          currentPath === dep ||
          new URL(currentPath, importerUrl).href === dep

// given that currentPath is absolute is equivalent to
          currentPath === dep ||
          currentPath === dep

// which is equivalent to
          currentPath === dep
```

Tal500 replied:
Tnx, just used your code instead.

bgoscinski commented on packages/vite/src/node/plugins/importAnalysisBuild.ts:
```diff
@@ -73,30 +73,66 @@ function preload(
+      for (let i = 0; i < possibleLinks.length; ++i) {
+        const currentPath = possibleLinks[i].href
+        if (
+          currentPath === dep ||
```
This does fix #9967 but breaks the same thing for absolute `base` haha!

### How it's broken
When `base` is relative then `dep` gets resolved to absolute URL by the `assetsURL` function and this comparison passes. However when `base` is absolute then `importerUrl` is not passed into `preload` function and `assetsURL` is [only prepending `dep` with `base` value](https://github.com/vitejs/vite/blob/8f68cdfb6650ca22c8128171b37036a595c414b5/packages/vite/src/node/plugins/importAnalysisBuild.ts#L156) which gives us a root-relative URL (like `/assets/foo.abc123.js`. This makes this comparison like `'http://example.com/assets/foo.abc123.js' === '/assets/foo.abc123.js'` which is `false`. 

### How to fix it
To fix this you'd need to resolve `dep` to absolute URL first. This could be done by running it through `new URL(dep, location.href)` first.

Tal500 replied:
> This does fix #9967 but breaks the same thing for absolute `base` haha!
> ### How it's broken
> 
> When `base` is relative then `dep` gets resolved to absolute URL by the `assetsURL` function and this comparison passes. However when `base` is absolute then `importerUrl` is not passed into `preload` function and `assetsURL` is [only prepending `dep` with `base` value](https://github.com/vitejs/vite/blob/8f68cdfb6650ca22c8128171b37036a595c414b5/packages/vite/src/node/plugins/importAnalysisBuild.ts#L156) which gives us a root-relative URL (like `/assets/foo.abc123.js`. This makes this comparison like `'http://example.com/assets/foo.abc123.js' === '/assets/foo.abc123.js'` which is `false`.
> ### How to fix it
> 
> To fix this we'd need to resolve `dep` to absolute URL before comparing it to `currentPath`. Ideally we'd do `new URL(dep, importerUrl)` but unfortunately for absolute `base` `importerUrl === undefined`. We can workaround this by using `location.href` instead: `new URL(dep, location.href)`
> 
> Also, feel free to use the test for this behavior from [bgoscinski@54a4d5d](https://github.com/bgoscinski/vite/commit/54a4d5dba587a42adef7f5e2416e245f8c625822)

I'll fix the code to match yours, but want to give you the credit for the test cases, so I'll wait until your PR will be accepted!

Tal500 committed:
fix: copy the suggested code from #9970

Tal500 committed:
chore: fix typo

Tal500 commented:
Now I think it's ready.
Suggest to merge #9970 first, both for credits and test units.

Tal500 committed:
chore: omit the fix for double stylesheets loading

The reason is that it's already fixed by #9970, and we want to make this PR independent.

Tal500 commented:
I had removed the fix of #9967 from this PR, since it's handled better in #9970.
Now both #9970 and this PR are independent, and merging one won't affect the other.

Note: Need to still implement in the future the suggestion appears in the discussion of https://github.com/vitejs/vite/pull/9970#discussion_r969505865, but this can be done later in a different PR, as this suggestion is nice to have.

So now I'm ready for review&merge.

Tal500 committed:
Merge branch 'main' into preload-on-system-format

Tal500 committed:
Merge branch 'vitejs:main' into preload-on-system-format

Tal500 committed:
Merge branch 'main' into preload-on-system-format

Tal500 committed:
chore: Merge branch 'main' into preload-on-system-format

Tal500 committed:
fix: return css inlining only for library mode

Tal500 committed:
Merge branch 'main' into preload-on-system-format

sapphi-red commented:
This implementation depends on regexs very heavily and I felt a bit fragile.
So I have a suggestion, how about using [`instantiate` hook](https://github.com/systemjs/systemjs/blob/main/docs/hooks.md#instantiateurl-parenturl---promise) of SystemJS? In this way, I think we could avoid regexs.

In entry chunks:
```js
if (!window.__viteLegacyDeps) {
  window.__viteLegacyDeps = {}

  const originalInstantiate = System.constructor.prototype.instantiate
  System.constructor.prototype.instantiate = function (args) {
    // preload using the information of `window.__viteLegacyDeps`
    return existingHook.call(this, args)
  }
}
```

In each chunks:
```js
window.__viteLegacyDeps['chunk file'] = [/* dep list */]
```

Tal500 commented:
> This implementation depends on regexs very heavily and I felt a bit fragile. 

This is why I added testing snapshots😃
I agree however about the principle. ESModules import are scanned well by Vite. Do you think we can somehow generate a map at the `renderChunk` stage(or such) between the `es`-formatted chunk to their deuended dynamically imported (other) chunks, and then use it directly for writing the dependency list?

> 
> In each chunks:
> 
> ```js
> window.__viteLegacyDeps['chunk file'] = [/* dep list */]
> ```

How do we get the dependency list here? Feels like we need to have the regexps (or an alternative solution) for this again.

sapphi-red commented:
> Do you think we can somehow generate a map at the `transform`/`renderChunk` stage(or such) between the `es`-formatted chunk to their (well-)detected dynamically imported (other) chunks, and then use it directly for writing the dependency list?

I think we can't. Dependency graph can only be obtained in `renderChunk`/`generateBundle`, but in those hooks the code is already transformed to SystemJS.

> How do we get the dependency list here? Feels like we need to have the regexps (or an alternative solution) for this again.

I guess we can use `dynamicImports` property of `ChunkInfo` to get what is dynamic-imported. Tracing static-imports  could be done [like we do now](https://github.com/vitejs/vite/blob/4b8a58703055a0eb5d8ddb04ef9f1eb081fbbc6a/packages/vite/src/node/plugins/importAnalysisBuild.ts#L508-L543).
https://rollupjs.org/guide/en/#generatebundle

Tal500 committed:
chore: Merge branch 'main' into preload-on-system-format

Tal500 committed:
fix: give empty list of systemJS dynamic import, if the `moduleParam` wasn't found well

Tal500 commented:
> > Do you think we can somehow generate a map at the `transform`/`renderChunk` stage(or such) between the `es`-formatted chunk to their (well-)detected dynamically imported (other) chunks, and then use it directly for writing the dependency list?
> 
> I think we can't. Dependency graph can only be obtained in `renderChunk`/`generateBundle`, but in those hooks the code is already transformed to SystemJS.
> 
> > How do we get the dependency list here? Feels like we need to have the regexps (or an alternative solution) for this again.
> 
> I guess we can use `dynamicImports` property of `ChunkInfo` to get what is dynamic-imported. Tracing static-imports could be done [like we do now](https://github.com/vitejs/vite/blob/4b8a58703055a0eb5d8ddb04ef9f1eb081fbbc6a/packages/vite/src/node/plugins/importAnalysisBuild.ts#L508-L543). https://rollupjs.org/guide/en/#generatebundle

@sapphi-red, continuing the logic in my previous comment, do you think that any (breaking?) change of Vite 4.0 can help us here?

Tal500 committed:
chore: merge branch 'main' into preload-on-system-format

sapphi-red commented:
> I deepen into Rollup chunk information, but it seems to tell us only about the dependencies sources, but not on the location on the chunk that the "import" is used, and `parseImports`/`parseImportsSystemJS` give us this extra information.

We only need that what will be dynamically imported in that chunk. I'm suggesting to declare the following code **in the head** of the chunk, not aside of the `import`.
```js
window.__viteLegacyDeps['chunk file'] = [/* dep list */]
```

For example, transform this code
```js
// main.000.js ------------------------------
System.register([], function (exports, module) {
  module.import('foo.111.js')
  module.import('bar.222.js')
});

// foo.111.js ------------------------------
System.register([], function (exports, module) {
  console.log('foo.111.js')
});

// foo.111.css ------------------------------
/* css content */

// bar.222.js ------------------------------
System.register([], function (exports, module) {
  console.log('bar.222.js')
  module.import('foobar.333.js')
});

// foobar.333.js ------------------------------
System.register([], function (exports, module) {
  console.log('fioobar.333.js')
});
```
to
```js
// main.000.js ------------------------------
if (!window.__viteLegacyDeps) {
  window.__viteLegacyDeps = {}

  const originalInstantiate = System.constructor.prototype.instantiate
  System.constructor.prototype.instantiate = function (args) {
    // preload using the information of `window.__viteLegacyDeps`
    return existingHook.call(this, args)
  }
}

window.__viteLegacyDeps['foo.111.js'] = ['foo.111.css']
window.__viteLegacyDeps['bar.222.js'] = []

System.register([], function (exports, module) {
  module.import('foo.111.js')
  module.import('bar.222.js')
});

// foo.111.js ------------------------------
System.register([], function (exports, module) {
  console.log('foo.111.js')
});

// foo.111.css ------------------------------
/* css content */

// bar.222.js ------------------------------
window.__viteLegacyDeps['foobar.333.js'] = []

System.register([], function (exports, module) {
  console.log('bar.222.js')
  module.import('foobar.333.js')
});

// foobar.333.js ------------------------------
System.register([], function (exports, module) {
  console.log('fioobar.333.js')
});
```

> How then can we overcome the removing process of pure css imports?

How about injecting `window.__viteLegacyPureCss = [/* pure css deps */]` and skipping them in `instantiate` when preloading?

> continuing the logic in my previous comment, do you think that any (breaking?) change of Vite 4.0 can help us here?

I'm not able to track the changes deeply but I guess no.

Tal500 committed:
refactor: cancel many regexp analysis, and use SystemJS `instantiate` instead

Tal500 commented:
Took some work but finally made it with SystemJS hooks!
Ready for review @sapphi-red :-)

Tal500 committed:
chore: merge branch 'main' into preload-on-system-format

Tal500 committed:
Merge branch 'main' into preload-on-system-format

Tal500 committed:
chore: fix the `@ts-ignore` issue on CI

Tal500 committed:
Merge branch 'main' into preload-on-system-format

Tal500 committed:
chore: add the non-nullish ! to "chunk.viteMetadata"

sapphi-red committed:
Merge branch 'main' into pr/9920

sapphi-red committed:
Merge branch 'main' into pr/9920

sapphi-red committed:
fix: inline CSS in non-lib mode if css.codeSplit is false

sapphi-red committed:
refactor: revert rename preloadMethod

sapphi-red committed:
test: test nested dynamic import

sapphi-red commented:
Sorry that it took a long time.

I tried this PR out and found that it doesn't work when there's a dynamic import in a dynamic imported chunk.
For example, in https://github.com/vitejs/vite/pull/9920/commits/211f4a7ca9e1659683957a5ae82ee4bef2499dff, "async2" should be blue but it isn't in legacy.
I guess this is because this PR's implementation only injects a single `__VITE_PRELOAD_LIST__`. `__VITE_PRELOAD_LIST__` has to exist in every chunk.

Tal500 commented:
> I'm sorry that it took a long time.
> 
> I tried this PR out and found that it doesn't work when there's a dynamic import in a dynamic imported chunk. For example, in [211f4a7](https://github.com/vitejs/vite/commit/211f4a7ca9e1659683957a5ae82ee4bef2499dff), "async2" should be blue but it isn't in legacy. I guess this is because this PR's implementation only injects a single `__VITE_PRELOAD_LIST__`. `__VITE_PRELOAD_LIST__` has to exist in every chunk.

What is the way to go with that?
I also wasn't sure about the global declaration policy in Vite, and the chunk loading.

sapphi-red commented:
> What is the way to go with that?

I think injecting the following code to every chunk in `generateBundle` hook would work.
```js
${preloadListMarker}.forEach((data) => {
  ${preloadLegacyImportsToken}[assetsURL(data[0], import.meta.url)] = {
    deps: data[1],
    pure: data.length === 3
  }
})
```

> I also wasn't sure about the global declaration policy in Vite, and the chunk loading.

This PR already depends on `window.__viteLegacyImports`. What do you mean by "global declaration policy"?
